src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -257,29 +246,40 @@ private void SetInferredTypes(ArrayBuilder<DeconstructionVariable> variables, Im
                 var variable = variables[i];
                 if (!variable.HasNestedVariables)
                 {
-                    switch (variable.Single.Kind)
+                    var pending = variable.Single;
+                    if ((object)pending.Type != null)
                     {
-                        case BoundKind.DeconstructionVariablePendingInference:
-                            {
-                                var pending = (DeconstructionVariablePendingInference)variable.Single;
-                                BoundExpression local = pending.SetInferredType(foundTypes[i], this, diagnostics);
-                                variables[i] = new DeconstructionVariable(local, local.Syntax);
-                            }
-                            break;
-                        case BoundKind.DiscardedExpression:
-                            {
-                                var pending = (BoundDiscardedExpression)variable.Single;
-                                if ((object)pending.Type == null)
-                                {
-                                    variables[i] = new DeconstructionVariable(pending.SetInferredType(foundTypes[i]), pending.Syntax);
-                                }
-                            }
-                            break;
+                        continue;
                     }
+
+                    variables[i] = new DeconstructionVariable(SetInferredType(pending, foundTypes[i], diagnostics), pending.Syntax);
                 }
             }
         }
 
+        private BoundExpression SetInferredType(BoundExpression expression, TypeSymbol type, DiagnosticBag diagnostics)
+        {
+            switch (expression.Kind)
+            {
+                case BoundKind.DeconstructionVariablePendingInference:
+                    {
+                        var pending = (DeconstructionVariablePendingInference)expression;
+                        return pending.SetInferredType(type, this, diagnostics);
+                    }
+                case BoundKind.DiscardedExpression:
+                    {
+                        var pending = (BoundDiscardedExpression)expression;
+                        if ((object)pending.Type == null)
+                        {
```

Comment:
Since the caller has already checked this, I will remove the `if` and add an assertion that the type is `null`.